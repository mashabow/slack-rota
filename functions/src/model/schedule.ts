import roundToNearestMinutes from "date-fns/roundToNearestMinutes";
import utcToZonedTime from "date-fns-tz/utcToZonedTime";
import { NonFunctionProperties } from "../util";

export const DAY_STRINGS = ["日", "月", "火", "水", "木", "金", "土"] as const;
export const INTERVAL_MINUTES = 5;

type ScheduleArgs = NonFunctionProperties<Schedule>;
export type ScheduleJSON = ScheduleArgs;

export class Schedule {
  // JST での時間
  readonly days: readonly number[]; // 0, 1, …, 6。日曜始まり
  readonly hour: number; // 0, 1, …, 23
  readonly minute: number; // 0, 5, …, 55

  private constructor(args: ScheduleArgs) {
    this.days = args.days;
    this.hour = args.hour;
    this.minute = args.minute;
  }

  toJSON(): ScheduleJSON {
    return {
      days: this.days,
      hour: this.hour,
      minute: this.minute,
    };
  }

  static fromJSON(json: ScheduleJSON): Schedule {
    return new Schedule({
      days: json.days,
      hour: json.hour,
      minute: json.minute,
    });
  }

  static dateToNearestSchedule(date: Date): Schedule {
    const utc = roundToNearestMinutes(date, {
      nearestTo: INTERVAL_MINUTES,
    });
    const jst = utcToZonedTime(utc, "Asia/Tokyo");

    return Schedule.fromJSON({
      days: [jst.getDay()],
      hour: jst.getHours(),
      minute: jst.getMinutes(),
    });
  }

  toString(): string {
    const days = this.days.map((day) => `${DAY_STRINGS[day]}曜`).join("・");
    const time = `${this.hour}:${this.minute.toString().padStart(2, "0")}`;
    return `${days}の ${time}`;
  }
}
