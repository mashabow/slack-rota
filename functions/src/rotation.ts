import { NonFunctionProperties, Optional } from "./util";

export const INTERVAL_MINUTES = 5;

type RotationJSON = NonFunctionProperties<Rotation>;

export class Rotation {
  readonly id: string;
  readonly members: readonly string[];
  readonly onDuty: string; // 担当者の user_id。member の中の 1 人
  readonly message: string;
  readonly channel: string;
  // JST
  readonly hour: number; // 0, 1, …, 23
  readonly minute: number; // 0, 5, …, 55

  private constructor(args: RotationJSON) {
    this.id = args.id;
    this.members = args.members;
    this.onDuty = args.onDuty;
    this.message = args.message;
    this.channel = args.channel;
    this.hour = args.hour;
    this.minute = args.minute;
  }

  toJSON(): RotationJSON {
    return {
      id: this.id,
      members: this.members,
      onDuty: this.onDuty,
      message: this.message,
      channel: this.channel,
      hour: this.hour,
      minute: this.minute,
    };
  }

  static fromJSON(json: Optional<RotationJSON, "id" | "onDuty">): Rotation {
    return new Rotation({
      id: json.id ?? new Date().valueOf().toString(),
      members: json.members,
      onDuty: json.onDuty ?? json.members[json.members.length - 1],
      message: json.message,
      channel: json.channel,
      hour: json.hour,
      minute: json.minute,
    });
  }

  rotate(): Rotation {
    const index = this.members.indexOf(this.onDuty);
    return new Rotation({
      ...this,
      onDuty: this.members[index + 1] ?? this.members[0],
    });
  }

  getOrderedRestMembers(): readonly string[] {
    const index = this.members.indexOf(this.onDuty);
    return [...this.members.slice(index + 1), ...this.members.slice(0, index)];
  }
}
