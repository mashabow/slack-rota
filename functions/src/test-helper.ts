import crypto from "crypto";
import querystring from "querystring";
import { HttpsFunction } from "firebase-functions";
import request from "supertest";

export const CONFIG = {
  slack: {
    bot_token: "dummy-bot-token",
    signing_secret: "dummy-signing-secret",
  },
};

export const postSlackEvent = (
  httpsFunction: HttpsFunction,
  body: Record<string, string>
): request.Test => {
  const encodedBody = querystring.stringify(body);

  // https://api.slack.com/authentication/verifying-requests-from-slack#step-by-step_walk-through_for_validating_a_request
  const timestamp = Math.floor(Date.now() / 1000);
  const hmac = crypto.createHmac("sha256", CONFIG.slack.signing_secret);
  hmac.update(`v0:${timestamp}:${encodedBody}`);
  const signature = `v0=${hmac.digest("hex")}`;

  return request(httpsFunction)
    .post("/events")
    .set({
      "x-slack-signature": signature,
      "x-slack-request-timestamp": timestamp.toString(),
    })
    .send(encodedBody);
};
