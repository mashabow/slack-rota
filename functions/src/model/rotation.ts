import { NonFunctionProperties, Optional } from "../util";
import { Schedule, ScheduleJSON } from "./schedule";

type RotationArgs = NonFunctionProperties<Rotation>;
type RotationJSON = Omit<RotationArgs, "schedule"> & {
  readonly schedule: ScheduleJSON;
};

export class Rotation {
  readonly id: string;
  readonly members: readonly string[];
  readonly onDuty: string; // 担当者の user_id。members の中の 1 人
  readonly message: string;
  readonly channel: string;
  readonly schedule: Schedule;

  private constructor(args: RotationArgs) {
    this.id = args.id;
    this.members = args.members;
    this.onDuty = args.onDuty;
    this.message = args.message;
    this.channel = args.channel;
    this.schedule = args.schedule;
  }

  toJSON(): RotationJSON {
    return {
      id: this.id,
      members: this.members,
      onDuty: this.onDuty,
      message: this.message,
      channel: this.channel,
      schedule: this.schedule.toJSON(),
    };
  }

  static fromJSON(json: Optional<RotationJSON, "id" | "onDuty">): Rotation {
    return new Rotation({
      id: json.id ?? new Date().valueOf().toString(),
      members: json.members,
      onDuty: json.onDuty ?? json.members[json.members.length - 1],
      message: json.message,
      channel: json.channel,
      schedule: Schedule.fromJSON(json.schedule),
    });
  }

  rotate(): Rotation {
    const index = this.members.indexOf(this.onDuty);
    return new Rotation({
      ...this,
      onDuty: this.members[index + 1] ?? this.members[0],
    });
  }

  getOrderedRestMembers(): readonly string[] {
    const index = this.members.indexOf(this.onDuty);
    return [...this.members.slice(index + 1), ...this.members.slice(0, index)];
  }
}
