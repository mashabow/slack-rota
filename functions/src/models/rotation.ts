import { NonFunctionProperties, Optional } from "../util";
import { Schedule, ScheduleJSON } from "./schedule";

type RotationArgs = NonFunctionProperties<Rotation>;
export type RotationJSON = Omit<RotationArgs, "schedule"> & {
  readonly schedule: ScheduleJSON;
};

export class Rotation {
  readonly id: string;
  readonly installationId: string;
  // members の先頭が担当者。store には、最後に post したときの状態で保存する
  readonly members: readonly string[];
  readonly message: string;
  readonly channel: string;
  readonly schedule: Schedule;
  readonly mentionAll: boolean;

  private constructor(args: RotationArgs) {
    this.id = args.id;
    this.installationId = args.installationId;
    this.members = args.members;
    this.message = args.message;
    this.channel = args.channel;
    this.schedule = args.schedule;
    this.mentionAll = args.mentionAll;
  }

  toJSON(): RotationJSON {
    return {
      id: this.id,
      installationId: this.installationId,
      members: this.members,
      message: this.message,
      channel: this.channel,
      schedule: this.schedule.toJSON(),
      mentionAll: this.mentionAll,
    };
  }

  static fromJSON(json: Optional<RotationJSON, "id">): Rotation {
    return new Rotation({
      id: json.id ?? Date.now().toString(),
      installationId: json.installationId,
      members: json.members,
      message: json.message,
      channel: json.channel,
      schedule: Schedule.fromJSON(json.schedule),
      mentionAll: json.mentionAll,
    });
  }

  rotate(): Rotation {
    return new Rotation({
      ...this,
      members: [...this.members.slice(1), this.members[0]],
    });
  }

  unrotate(): Rotation {
    return new Rotation({
      ...this,
      members: [
        this.members[this.members.length - 1],
        ...this.members.slice(0, -1),
      ],
    });
  }
}
