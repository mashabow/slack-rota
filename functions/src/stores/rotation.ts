import { Rotation } from "../models/rotation";
import { Schedule } from "../models/schedule";

export class RotationStore {
  private collection: FirebaseFirestore.CollectionReference;

  constructor(db: FirebaseFirestore.Firestore) {
    this.collection = db.collection("rotations");
  }

  async set(rotation: Rotation): Promise<void> {
    await this.collection.doc(rotation.id).set(rotation.toJSON());
  }

  async get(rotationId: string): Promise<Rotation | null> {
    const doc = await this.collection.doc(rotationId).get();
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return doc.exists ? Rotation.fromJSON(doc.data() as any) : null;
  }

  async getAllBySchedule(schedule: Schedule): Promise<readonly Rotation[]> {
    if (schedule.days.length !== 1) {
      throw new Error("Length of schedule.days must be 1");
    }

    const snapshot = await this.collection
      .where("schedule.days", "array-contains", schedule.days[0])
      .where("schedule.hour", "==", schedule.hour)
      .where("schedule.minute", "==", schedule.minute)
      .get();
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return snapshot.docs.map((doc) => Rotation.fromJSON(doc.data() as any));
  }

  async getAllByMember(member: string): Promise<readonly Rotation[]> {
    const snapshot = await this.collection
      .where("members", "array-contains", member)
      .get();
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return snapshot.docs.map((doc) => Rotation.fromJSON(doc.data() as any));
  }

  async delete(rotationId: string): Promise<void> {
    await this.collection.doc(rotationId).delete();
  }
}
