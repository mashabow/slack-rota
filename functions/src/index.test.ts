import { describe, it, expect } from "@jest/globals";
import {
  MockedWebClient,
  MockWebClient,
} from "@slack-wrench/jest-mock-web-client";
import dotenv from "dotenv";
import * as admin from "firebase-admin";
import functionsTest from "firebase-functions-test";
import { RotationJSON } from "./model/rotation";
import { CONFIG, postSlackEvent } from "./test-helper";

dotenv.config({ path: `${__dirname}/../../.env` });
const { TEST_PROJECT_ID } = process.env;
if (!TEST_PROJECT_ID) {
  throw new Error("Environment variable TEST_PROJECT_ID not set.");
}

const test = functionsTest(
  {
    projectId: `${TEST_PROJECT_ID}`,
    databaseURL: `https://${TEST_PROJECT_ID}.firebaseio.com`,
    storageBucket: `${TEST_PROJECT_ID}.appspot.com`,
  },
  `${__dirname}/../../serviceAccountKey.json`
);
test.mockConfig(CONFIG);

import { slack, cron } from "./index";

describe("functions", () => {
  afterAll(() => {
    test.cleanup();
  });

  // Mock Slack API client
  let client: MockWebClient;
  beforeEach(async () => {
    client = MockedWebClient.mock.instances[0];
  });
  afterEach(async () => {
    client.chat.postMessage.mockClear();
  });

  describe("slack", () => {
    describe("/rota command", () => {
      it("opens SettingModal", async () => {
        const res = await postSlackEvent(slack, {
          command: "/rota",
          text: "foobar",
          channel_id: "channel-id",
          trigger_id: "trigger-id",
          // ...more unused parameters
        });
        expect(res.body).toEqual({}); // ack
        expect(client.views.open.mock.calls).toMatchSnapshot();
      });
    });
  });

  describe("cron", () => {
    const wrappedCron = test.wrap(cron);
    const rotationsRef = admin.firestore().collection("rotations");
    const rotations: readonly RotationJSON[] = [
      {
        id: "rotation-1",
        members: ["user-a", "user-b", "user-c"],
        onDuty: "user-a",
        message: "rotation-1 message",
        channel: "channel-1",
        schedule: {
          days: [0],
          hour: 7,
          minute: 35,
        },
      },
      {
        id: "rotation-2",
        members: ["user-p", "user-q"],
        onDuty: "user-q",
        message: "rotation-2 message",
        channel: "channel-2",
        schedule: {
          days: [0, 6],
          hour: 7,
          minute: 35,
        },
      },
      {
        id: "rotation-3",
        members: ["user-s", "user-t"],
        onDuty: "user-s",
        message: "rotation-3 message",
        channel: "channel-3",
        schedule: {
          days: [0],
          hour: 22,
          minute: 35,
        },
      },
      {
        id: "rotation-4",
        members: ["user-x", "user-y", "user-z"],
        onDuty: "user-y",
        message: "rotation-4 message",
        channel: "channel-4",
        schedule: {
          days: [2, 3, 4, 5],
          hour: 7,
          minute: 35,
        },
      },
    ];

    beforeEach(async () => {
      // Prepare rotations in Firestore
      await Promise.all(
        rotations.map((rotation) => rotationsRef.doc(rotation.id).set(rotation))
      );
    });
    afterEach(async () => {
      // Delete all rotations from Firestore
      const snapshot = await rotationsRef.get();
      await Promise.all(snapshot.docs.map((doc) => doc.ref.delete()));
    });

    it("posts matched rotations and updates onDuty fields of them", async () => {
      await wrappedCron({
        timestamp: "2020-08-08T22:33:44.000Z", // Sun, 09 Aug 2020 07:33:44 JST
      });
      expect(client.chat.postMessage.mock.calls).toMatchSnapshot();
      expect((await rotationsRef.get()).docs.map((doc) => doc.data())).toEqual([
        { ...rotations[0], onDuty: "user-b" },
        { ...rotations[1], onDuty: "user-p" },
        rotations[2],
        rotations[3],
      ]);
    });

    it("does nothing when no rotation matched", async () => {
      await wrappedCron({
        timestamp: "2020-08-09T22:33:44.000Z", // Mon, 10 Aug 2020 07:33:44 JST
      });
      expect(client.chat.postMessage.mock.calls).toEqual([]);
      expect((await rotationsRef.get()).docs.map((doc) => doc.data())).toEqual(
        rotations
      );
    });
  });
});
