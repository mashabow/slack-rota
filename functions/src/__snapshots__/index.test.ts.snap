// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`functions cron posts matched rotations and updates onDuty fields of them 1`] = `
Array [
  Array [
    Object {
      "blocks": Array [
        Object {
          "text": Object {
            "text": "rotation-1 message
",
            "type": "mrkdwn",
            "verbatim": false,
          },
          "type": "section",
        },
        Object {
          "accessory": Object {
            "action_id": "overflow_menu",
            "options": Array [
              Object {
                "text": Object {
                  "emoji": true,
                  "text": "ÂâäÈô§",
                  "type": "plain_text",
                },
                "value": "delete:rotation-1",
              },
            ],
            "type": "overflow",
          },
          "text": Object {
            "text": "üëë <@user-a|>‚Üí <@user-b|>‚Üí<@user-c|>",
            "type": "mrkdwn",
            "verbatim": true,
          },
          "type": "section",
        },
      ],
      "channel": "channel-1",
      "text": "rotation-1 message",
      "token": "dummy-bot-token",
    },
  ],
  Array [
    Object {
      "blocks": Array [
        Object {
          "text": Object {
            "text": "rotation-2 message
",
            "type": "mrkdwn",
            "verbatim": false,
          },
          "type": "section",
        },
        Object {
          "accessory": Object {
            "action_id": "overflow_menu",
            "options": Array [
              Object {
                "text": Object {
                  "emoji": true,
                  "text": "ÂâäÈô§",
                  "type": "plain_text",
                },
                "value": "delete:rotation-2",
              },
            ],
            "type": "overflow",
          },
          "text": Object {
            "text": "üëë <@user-q|>‚Üí<@user-p|>",
            "type": "mrkdwn",
            "verbatim": true,
          },
          "type": "section",
        },
      ],
      "channel": "channel-2",
      "text": "rotation-2 message",
      "token": "dummy-bot-token",
    },
  ],
]
`;

exports[`functions slack /rota command opens SettingModal 1`] = `
Array [
  Array [
    Object {
      "token": "dummy-bot-token",
      "trigger_id": "trigger-id",
      "view": Object {
        "blocks": Array [
          Object {
            "block_id": "members",
            "element": Object {
              "action_id": "members",
              "type": "multi_users_select",
            },
            "label": Object {
              "emoji": true,
              "text": "„É°„É≥„Éê„Éº",
              "type": "plain_text",
            },
            "optional": false,
            "type": "input",
          },
          Object {
            "block_id": "message",
            "element": Object {
              "action_id": "message",
              "multiline": true,
              "type": "plain_text_input",
            },
            "label": Object {
              "emoji": true,
              "text": "„É°„ÉÉ„Çª„Éº„Ç∏",
              "type": "plain_text",
            },
            "optional": false,
            "type": "input",
          },
          Object {
            "block_id": "days",
            "element": Object {
              "action_id": "days",
              "options": Array [
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "Êó•Êõú",
                    "type": "plain_text",
                  },
                  "value": "0",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "ÊúàÊõú",
                    "type": "plain_text",
                  },
                  "value": "1",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "ÁÅ´Êõú",
                    "type": "plain_text",
                  },
                  "value": "2",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "Ê∞¥Êõú",
                    "type": "plain_text",
                  },
                  "value": "3",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "Êú®Êõú",
                    "type": "plain_text",
                  },
                  "value": "4",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "ÈáëÊõú",
                    "type": "plain_text",
                  },
                  "value": "5",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "ÂúüÊõú",
                    "type": "plain_text",
                  },
                  "value": "6",
                },
              ],
              "type": "multi_static_select",
            },
            "label": Object {
              "emoji": true,
              "text": "ÊõúÊó•",
              "type": "plain_text",
            },
            "optional": false,
            "type": "input",
          },
          Object {
            "block_id": "hour",
            "element": Object {
              "action_id": "hour",
              "initial_option": Object {
                "text": Object {
                  "emoji": true,
                  "text": "10ÊôÇ",
                  "type": "plain_text",
                },
                "value": "10",
              },
              "options": Array [
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "0ÊôÇ",
                    "type": "plain_text",
                  },
                  "value": "0",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "1ÊôÇ",
                    "type": "plain_text",
                  },
                  "value": "1",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "2ÊôÇ",
                    "type": "plain_text",
                  },
                  "value": "2",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "3ÊôÇ",
                    "type": "plain_text",
                  },
                  "value": "3",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "4ÊôÇ",
                    "type": "plain_text",
                  },
                  "value": "4",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "5ÊôÇ",
                    "type": "plain_text",
                  },
                  "value": "5",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "6ÊôÇ",
                    "type": "plain_text",
                  },
                  "value": "6",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "7ÊôÇ",
                    "type": "plain_text",
                  },
                  "value": "7",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "8ÊôÇ",
                    "type": "plain_text",
                  },
                  "value": "8",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "9ÊôÇ",
                    "type": "plain_text",
                  },
                  "value": "9",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "10ÊôÇ",
                    "type": "plain_text",
                  },
                  "value": "10",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "11ÊôÇ",
                    "type": "plain_text",
                  },
                  "value": "11",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "12ÊôÇ",
                    "type": "plain_text",
                  },
                  "value": "12",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "13ÊôÇ",
                    "type": "plain_text",
                  },
                  "value": "13",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "14ÊôÇ",
                    "type": "plain_text",
                  },
                  "value": "14",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "15ÊôÇ",
                    "type": "plain_text",
                  },
                  "value": "15",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "16ÊôÇ",
                    "type": "plain_text",
                  },
                  "value": "16",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "17ÊôÇ",
                    "type": "plain_text",
                  },
                  "value": "17",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "18ÊôÇ",
                    "type": "plain_text",
                  },
                  "value": "18",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "19ÊôÇ",
                    "type": "plain_text",
                  },
                  "value": "19",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "20ÊôÇ",
                    "type": "plain_text",
                  },
                  "value": "20",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "21ÊôÇ",
                    "type": "plain_text",
                  },
                  "value": "21",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "22ÊôÇ",
                    "type": "plain_text",
                  },
                  "value": "22",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "23ÊôÇ",
                    "type": "plain_text",
                  },
                  "value": "23",
                },
              ],
              "type": "static_select",
            },
            "label": Object {
              "emoji": true,
              "text": "ÊôÇ",
              "type": "plain_text",
            },
            "optional": false,
            "type": "input",
          },
          Object {
            "block_id": "minute",
            "element": Object {
              "action_id": "minute",
              "initial_option": Object {
                "text": Object {
                  "emoji": true,
                  "text": "0ÂàÜ",
                  "type": "plain_text",
                },
                "value": "0",
              },
              "options": Array [
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "0ÂàÜ",
                    "type": "plain_text",
                  },
                  "value": "0",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "5ÂàÜ",
                    "type": "plain_text",
                  },
                  "value": "5",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "10ÂàÜ",
                    "type": "plain_text",
                  },
                  "value": "10",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "15ÂàÜ",
                    "type": "plain_text",
                  },
                  "value": "15",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "20ÂàÜ",
                    "type": "plain_text",
                  },
                  "value": "20",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "25ÂàÜ",
                    "type": "plain_text",
                  },
                  "value": "25",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "30ÂàÜ",
                    "type": "plain_text",
                  },
                  "value": "30",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "35ÂàÜ",
                    "type": "plain_text",
                  },
                  "value": "35",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "40ÂàÜ",
                    "type": "plain_text",
                  },
                  "value": "40",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "45ÂàÜ",
                    "type": "plain_text",
                  },
                  "value": "45",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "50ÂàÜ",
                    "type": "plain_text",
                  },
                  "value": "50",
                },
                Object {
                  "text": Object {
                    "emoji": true,
                    "text": "55ÂàÜ",
                    "type": "plain_text",
                  },
                  "value": "55",
                },
              ],
              "type": "static_select",
            },
            "label": Object {
              "emoji": true,
              "text": "ÂàÜ",
              "type": "plain_text",
            },
            "optional": false,
            "type": "input",
          },
        ],
        "callback_id": "submit_callback",
        "close": Object {
          "emoji": true,
          "text": "„Ç≠„É£„É≥„Çª„É´",
          "type": "plain_text",
        },
        "private_metadata": "{\\"channel\\":\\"channel-id\\"}",
        "submit": Object {
          "emoji": true,
          "text": "Ë®≠ÂÆö„Åô„Çã",
          "type": "plain_text",
        },
        "title": Object {
          "emoji": true,
          "text": "Rota",
          "type": "plain_text",
        },
        "type": "modal",
      },
    },
  ],
]
`;

exports[`functions slack submit from SettingModal posts SettingSuccessMessage and creates a rotation on Firestore 1`] = `
Array [
  Array [
    Object {
      "blocks": Array [
        Object {
          "text": Object {
            "text": "<@user-id|>„Åï„Çì„Åå„É≠„Éº„ÉÜ„Éº„Ç∑„Éß„É≥„ÇíË®≠ÂÆö„Åó„Åæ„Åó„ÅüÔºÅ

ÊúàÊõú„ÉªÊ∞¥Êõú„ÉªÈáëÊõú„ÅÆ 23:45 „Å´ üëá „ÅÆ„Çà„ÅÜ„Å™ÊÑü„Åò„Åß„ÅäÁü•„Çâ„Åõ„Åó„Åæ„Åô",
            "type": "mrkdwn",
            "verbatim": true,
          },
          "type": "section",
        },
        Object {
          "text": Object {
            "text": "&gt; „Å¶„Åô„Å¶„Åô
&gt;
&gt; *„ÉÜ„Çπ„Éà*„Åß„Åô
",
            "type": "mrkdwn",
            "verbatim": false,
          },
          "type": "section",
        },
        Object {
          "accessory": Object {
            "action_id": "overflow_menu",
            "options": Array [
              Object {
                "text": Object {
                  "emoji": true,
                  "text": "ÂâäÈô§",
                  "type": "plain_text",
                },
                "value": "delete:1597200000000",
              },
            ],
            "type": "overflow",
          },
          "text": Object {
            "text": "&gt; üëë <@user-a|>‚Üí <@user-b|>‚Üí<@user-c|>
&gt; ",
            "type": "mrkdwn",
            "verbatim": true,
          },
          "type": "section",
        },
      ],
      "channel": "channel-id",
      "text": "<@user-id> „Åï„Çì„Åå„É≠„Éº„ÉÜ„Éº„Ç∑„Éß„É≥„Çí‰ΩúÊàê„Åó„Åæ„Åó„ÅüÔºÅ",
      "token": "dummy-bot-token",
    },
  ],
]
`;
